pip install jupyter
pip install elasticsearch
pip install 

## Download dataset

In [1]:
from datasets import load_dataset

# dataset link: https://huggingface.co/datasets/hugfaceguy0001/stanford_plato
dataset = load_dataset("hugfaceguy0001/stanford_plato", cache_dir="../data")

README.md:   0%|          | 0.00/2.49k [00:00<?, ?B/s]

(…)-00000-of-00001-65b5548e09cbc609.parquet:   0%|          | 0.00/90.0M [00:00<?, ?B/s]

Generating train split:   0%|          | 0/1776 [00:00<?, ? examples/s]

This dataset includes 1776 articles, each explaining one philosophy term/people/topic. It has 8 features:

* shorturl: The shorturl for the article. For example, the shorturl 'abduction' correspond to the page https://plato.stanford.edu/entries/abduction/
* title: The title of the article.
* pubinfo: The publication information.
* preamble: The preface text of the article. The data is a list, each item of the list is a paragraph of the data. I choose not to break the paragraph structure. Certainly, you can merge them by, for example, ''.join(data['preamble'])
toc: Table of contents. Also represented as list. Each item is a dictionary, the 'content_title' is the main content title, and the 'sub_toc' is a list of subcontent titles.
* main_text: The main text of the article. The data is also a list, each item represents a section of the article. Each item is a dictionary, 'section_title' is the title of the section, 'main_content' is a list of paragraphs before subsections, 'subsections' is a list of subsections, each item is also a dictionary, has its own title 'subsection_title' and list of paragraphs 'content'.
* bibliography: list of bibliography.
* related_entries: list of entries related to the current entry.

In [10]:
df = dataset['train'].to_pandas()
df.to_csv("../data/stanford_plato_train.csv", index=False)

In [9]:
df

shorturl                               title  \
0                     abduction                           Abduction   
1                       abelard                       Peter Abelard   
2                    abhidharma                          Abhidharma   
3                     abilities                           Abilities   
4                  abner-burgos                     Abner of Burgos   
...                         ...                                 ...   
1771     utilitarianism-history       The History of Utilitarianism   
1772  value-intrinsic-extrinsic       Intrinsic vs. Extrinsic Value   
1773            value-pluralism                     Value Pluralism   
1774             voting-methods                      Voting Methods   
1775           wyclif-political  John Wyclif’s Political Philosophy   

                                                pubinfo  \
0     First published Wed Mar 9, 2011; substantive r...   
1     First published Tue Aug 3, 2004; substantive r...   
2     First published Mon Aug 16, 2010; substantive ...   
3     First published Tue Jan 26, 2010; substantive ...   
4     First published Mon Jul 9, 2012; substantive r...   
...                                                 ...   
1771  First published Fri Mar 27, 2009; substantive ...   
1772  First published Tue Oct 22, 2002; substantive ...   
1773  First published Tue Jun 20, 2006; substantive ...   
1774  First published Wed Aug 3, 2011; substantive r...   
1775  First published Sat Jun 10, 2006; substantive ...   

                                               preamble  \
0     [\nIn the philosophical literature, the term “...   
1     [\nPeter Abelard (1079–21 April 1142) [‘Abaila...   
2     [\nThe first centuries after Śākyamuni Buddha’...   
3     [\nIn the accounts we give of one another, cla...   
4     [\nAbner of Burgos (Alfonso de Valladolid; c. ...   
...                                                 ...   
1771  [\n\nUtilitarianism is one of the most powerfu...   
1772  [\nIntrinsic value has traditionally been thou...   
1773  [\n\nThe word ‘pluralism’ generally refers to ...   
1774  [\nA fundamental problem faced by any group of...   
1775  [\n\nThe central idea of Wyclif's political ph...   

                                                    toc  \
0     [{'content_title': '1. Abduction: The General ...   
1     [{'content_title': '1. Life and Works', 'sub_t...   
2     [{'content_title': '1. Abhidharma: its origins...   
3     [{'content_title': '1. A taxonomy', 'sub_toc':...   
4     [{'content_title': '1. Life', 'sub_toc': []}, ...   
...                                                 ...   
1771  [{'content_title': '1. Precursors to the Class...   
1772  [{'content_title': '1. What Has Intrinsic Valu...   
1773  [{'content_title': '1. Some Preliminary Clarif...   
1774  [{'content_title': '1. The Problem: Who ', 'su...   
1775  [{'content_title': '1. Wyclif's Later Works ',...   

                                              main_text  \
0     [{'main_content': ['
You happen to know that T...   
1     [{'main_content': [], 'section_title': '1. Lif...   
2     [{'main_content': ['
The early history of Budd...   
3     [{'main_content': ['
What is an ability? On on...   
4     [{'main_content': ['
There are not many source...   
...                                                 ...   
1771  [{'main_content': ['

Though the first systema...   
1772  [{'main_content': ['
The question “What is int...   
1773  [{'main_content': [], 'section_title': '1. Som...   
1774  [{'main_content': ['
Suppose that there is a g...   
1775  [{'main_content': ['

Government and the relat...   

                                           bibliography  \
0     [Achinstein, P., 2001. The Book of Evidence, O...   
1     [Carmen ad Astralabium. Edited by J. M. A.\nRu...   
2     [[A] Aṅguttara-nikāya, [Dhs-a] Atthasālinī\n(D...   
3     [Albritton, Rogers, 1985. “Freedom of Will and...   
4     [Meyasher aqob, G. Gluskina (ed. 

In [75]:
example_article = dataset['train'][0]
example_article

{'shorturl': 'abduction',
 'title': 'Abduction',
 'pubinfo': 'First published Wed Mar 9, 2011; substantive revision Tue May 18, 2021',
 'preamble': ['\nIn the philosophical literature, the term “abduction” is\nused in two related but different senses. In both senses, the term\nrefers to some form of explanatory reasoning. However, in the\nhistorically first sense, it refers to the place of explanatory\nreasoning in generating hypotheses, while in the sense in\nwhich it is used most frequently in the modern literature it refers to\nthe place of explanatory reasoning in justifying hypotheses.\nIn the latter sense, abduction is also often called “Inference\nto the Best Explanation.”',
  '\nThis entry is exclusively concerned with abduction in the modern\nsense, although there is a supplement on abduction in the historical\nsense, which had its origin in the work of Charles Sanders\nPeirce—see the',
  '\nSee also the entry on\n scientific discovery,\n in particular the section on discovery

## Разбиваем на чанки и получаем метаданные

* Чтобы эффективно искать как по философским концепциям или по философам, нужно классифицировать статьи в датасете по категориям. Каждая статья должна быть помечена ключевыми терминами, философскими школами или конкретными философами. 
* Чтобы организовать правильную навигацию по статьям, можно включить эту информацию в поисковый индекс (добавлять заголовки в метаданные для быстрого поиска по секциям).

`TODO` для добавления в метаданных философских школ, можт понадобится поиск по словарю ("The Dictionary of Philosophy" или "The Cambridge Dictionary of Philosophy", Wikiledia: "Philosophical Schools"/"Philosophical Terms"). Т.к. это работает плохо: 
```python
elif ent.label_ in ["ORG", "NORP", "LANGUAGE"]:  # Organizations and religious/political groups
   concepts.add(ent.text)
```

In [16]:
import os
from datasets import load_dataset
from elasticsearch import Elasticsearch, helpers
import nltk
from nltk.tokenize import sent_tokenize
import spacy
import json
from tqdm import tqdm
from typing import List, Dict, Any
import hashlib

In [21]:
!python -m spacy download en_core_web_sm

  Using cached https://github.com/explosion/spacy-models/releases/download/en_core_web_sm-3.8.0/en_core_web_sm-3.8.0-py3-none-any.whl (12.8 MB)
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_sm')


In [22]:
nltk.download('punkt')

# Load spaCy model for named entity recognition
nlp = spacy.load("en_core_web_sm")

[nltk_data] Downloading package punkt to /home/codespace/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [69]:
class PhilosophyDataProcessor:
    def extract_entities(self, text: str) -> tuple:
        doc = nlp(text)
        philosophers = set()
        concepts = set()
        
        for ent in doc.ents:
            if ent.label_ == "PERSON":
                philosophers.add(ent.text)
            elif ent.label_ in ["ORG", "NORP", "LANGUAGE"]:  # Organizations and religious/political groups
                concepts.add(ent.text)
        
        return list(philosophers), list(concepts)

    def create_chunk_id(self, text: str, article_title: str, chunk_type: str) -> str:
        # Create a unique identifier for the chunk
        content = f"{text}{article_title}{chunk_type}"
        return hashlib.md5(content.encode()).hexdigest()

    def process_article(self, article: Dict[str, Any]) -> List[Dict[str, Any]]:
        chunks = []
        
        # Process preamble
        preamble_text = ' '.join(article['preamble'])
        philosophers, concepts = self.extract_entities(preamble_text)
        
        # Create full article chunk
        full_article_text = preamble_text
        for section in article['main_text']:
            section_text = ' '.join(section['main_content'])
            full_article_text += ' ' + section_text
        
        full_article_chunk = {
            'chunk_id': self.create_chunk_id(full_article_text, article['title'], 'full_article'),
            'article_title': article['title'],
            'section_path': [],
            'chunk_type': 'full_article',
            'philosophers': philosophers,
            'concepts': concepts,
            'content': full_article_text
        }
        chunks.append(full_article_chunk)
        
        # Process sections
        for section in article['main_text']:
            section_text = ' '.join(section['main_content'])
            section_philosophers, section_concepts = self.extract_entities(section_text)
            
            section_chunk = {
                'chunk_id': self.create_chunk_id(section_text, article['title'], 'section'),
                'article_title': article['title'],
                'section_path': [section['section_title']],
                'chunk_type': 'section',
                'philosophers': list(set(philosophers + section_philosophers)),
                'concepts': list(set(concepts + section_concepts)),
                'content': section_text
            }
            chunks.append(section_chunk)
            
            # Process paragraphs
            for paragraph in section['main_content']:
                if len(paragraph.split()) > 20:  
                    para_philosophers, para_concepts = self.extract_entities(paragraph)
                    
                    paragraph_chunk = {
                        'chunk_id': self.create_chunk_id(paragraph, article['title'], 'paragraph'),
                        'article_title': article['title'],
                        'section_path': [section['section_title']],
                        'chunk_type': 'paragraph',
                        'philosophers': list(set(philosophers + section_philosophers + para_philosophers)),
                        'concepts': list(set(concepts + section_concepts + para_concepts)),
                        'content': paragraph
                    }
                    chunks.append(paragraph_chunk)

            # Process subsections
            for subsection in section.get('subsections', []):
                subsection_text = ' '.join(subsection['content'])
                subsection_philosophers, subsection_concepts = self.extract_entities(subsection_text)
                
                subsection_chunk = {
                    'chunk_id': self.create_chunk_id(subsection_text, article['title'], 'subsection'),
                    'article_title': article['title'],
                    'section_path': [section['section_title'], subsection['subsection_title']],
                    'chunk_type': 'subsection',
                    'philosophers': list(set(philosophers + section_philosophers + subsection_philosophers)),
                    'concepts': list(set(concepts + section_concepts + subsection_concepts)),
                    'content': subsection_text
                }
                chunks.append(subsection_chunk)
        
        return chunks

In [70]:
# Initialize processor
processor = PhilosophyDataProcessor()

# Process one article as an example
example_article = dataset['train'][0]
chunks = processor.process_article(example_article)

In [72]:
df = pd.DataFrame(chunks)
df

chunk_id article_title  \
0   92dd2f6b5d033c12efb2f8ee3047601e     Abduction   
1   a3b5103db83e97338eb7270a86a63c65     Abduction   
2   81e573a9d329de10a9651c19c4907dc2     Abduction   
3   1738ba0482fa09f893136ec9e10ebff5     Abduction   
4   34bf01c649b729cd3d740f4d5a7cdadf     Abduction   
5   501b7b05cb58454829b17faa1ca0b67f     Abduction   
6   6c943762de83a112fe91d3de929913e8     Abduction   
7   806563744b1ad2c6fdc20c84694a9061     Abduction   
8   2e6cd6b7abb9bc8faf2847668c92552a     Abduction   
9   fb65974393a7620588b7c530d53e9082     Abduction   
10  8499745946def21e570ae45e9a8b2a82     Abduction   
11  1a750c640ab45d4dafc37540fd941b76     Abduction   
12  2bd101c6a5154c559d90b59afe196a2e     Abduction   
13  4f3f89d39f6e17a689d244cf3c1f8db4     Abduction   
14  1c4c184c0ddd0f96760e6fb93aa319fc     Abduction   
15  f7ecd1a32026e385eb2478522b11f166     Abduction   
16  10df8e48bc88212af4767ab5bb2cc362     Abduction   
17  169088bc3b24a5b29621bf52c9a4b680     Abduction   
18  7b955e4553c4bf8f4ff01e2f394ae4b5     Abduction   
19  a769e7233c923d5482ef84a18d34e322     Abduction   
20  c8927e679629bd82151236f34dcb5f8e     Abduction   
21  209998281eac41cd3e3b55194219adaa     Abduction   
22  bd8a142a8ae7d7250e4723f684d50b64     Abduction   
23  bb0345399a3ea255e6d31f230e818d53     Abduction   
24  c9249aa1fd1ac7dc1b7b71164a3de79d     Abduction   
25  bb09ab0b5a7f2dd68c1057691287d985     Abduction   
26  6ae1afd1662837a0294f1ac43b46c62c     Abduction   
27  69da7307a5a7eb89d8c73a877841e42e     Abduction   
28  603011c48c8bfc490ae806a1ba1b143c     Abduction   
29  9f6b5d5c458131ee8547cc2b36012efd     Abduction   
30  e429fb5c3e2a99a629105f9237b66a2d     Abduction   
31  d9e78e99955c0b0ae1dc121c3099336d     Abduction   
32  489106242e16717e1276835ca38b25f9     Abduction   
33  025cb383324ec21cd8cb553cf1a8f8f9     Abduction   
34  17ed98678b554b569ba01d9dce117a44     Abduction   
35  944b677a84a379e8805518f72011e74c     Abduction   
36  5238179758572338829937c7dda59f32     Abduction   
37  d5e38b1922a893a6d1b887c9ea7a6c27     Abduction   
38  5d84563da86fea01adfe1235cd360bdd     Abduction   
39  337d80c564b6a3ca35e08b80301661b8     Abduction   
40  2534a5a8ae83a361431b01c8e721d944     Abduction   

                                         section_path    chunk_type  \
0                                                  []  full_article   
1                    [1. Abduction: The General Idea]       section   
2                    [1. Abduction: The General Idea]     paragraph   
3                    [1. Abduction: The General Idea]     paragraph   
4                    [1. Abduction: The General Idea]     paragraph   
5                    [1. Abduction: The General Idea]     paragraph   
6   [1. Abduction: The General Idea, 1.1 Deduction...    subsection   
7   [1. Abduction: The General Idea, 1.2 The ubiqu...    subsection   
8                          [2. Explicating Abduction]       section   
9                          [2. Explicating Abduction]     paragraph   
10                         [2. Explicating Abduction]     paragraph   
11                         [2. Explicating Abduction]     paragraph   
12                         [2. Explicating Abduction]     paragraph   
13                         [2. Explicating Abduction]     paragraph   
14                         [2. Explicating Abduction]     paragraph   
15                         [2. Explicating Abduction]     paragraph   
16                         [2. Explicating Abduction]     paragraph   
17                         [2. Explicating Abduction]     paragraph   
18                         [2. Explicating Abduction]     paragraph   
19                         [2. Explicating Abduction]     paragraph   
20                         [2. Explicating Abduction]     paragraph   
21                         [2. Explicating Abduction]     paragraph   
22                         [2. Explicating Abduction]     paragraph   
23                         [2. Explicating Abduction]  

In [73]:
df.iloc[0].to_dict()

{'chunk_id': '92dd2f6b5d033c12efb2f8ee3047601e',
 'article_title': 'Abduction',
 'section_path': [],
 'chunk_type': 'full_article',
 'philosophers': ['Charles Sanders\nPeirce'],
 'concepts': ['Inference'],
 'content': '\nIn the philosophical literature, the term “abduction” is\nused in two related but different senses. In both senses, the term\nrefers to some form of explanatory reasoning. However, in the\nhistorically first sense, it refers to the place of explanatory\nreasoning in generating hypotheses, while in the sense in\nwhich it is used most frequently in the modern literature it refers to\nthe place of explanatory reasoning in justifying hypotheses.\nIn the latter sense, abduction is also often called “Inference\nto the Best Explanation.” \nThis entry is exclusively concerned with abduction in the modern\nsense, although there is a supplement on abduction in the historical\nsense, which had its origin in the work of Charles Sanders\nPeirce—see the \nSee also the entry on\n sci

In [74]:
df[df['content'].str.contains("What leads you to the conclusion", case=False, na=False)]

chunk_id article_title  \
0  92dd2f6b5d033c12efb2f8ee3047601e     Abduction   
1  a3b5103db83e97338eb7270a86a63c65     Abduction   
5  501b7b05cb58454829b17faa1ca0b67f     Abduction   

                       section_path    chunk_type  \
0                                []  full_article   
1  [1. Abduction: The General Idea]       section   
5  [1. Abduction: The General Idea]     paragraph   

                                        philosophers  \
0                          [Charles Sanders\nPeirce]   
1  [Harry, Tim, Churchill, Charles Sanders\nPeirc...   
5  [Harry, Tim, Churchill, Charles Sanders\nPeirc...   

                                            concepts  \
0                                        [Inference]   
1  [Winston\nChurchill, Hilary Putnam’s, Inferenc...   
5  [Winston\nChurchill, Hilary Putnam’s, Inferenc...   

                                             content  
0  \nIn the philosophical literature, the term “a...  
1  \nYou happen to know that Tim and Harry have r...  
5  \nIn these examples, the conclusions do not fo...

## Загрузка в ElasticSearch

Running ElasticSearch:
```
docker run -it \
    --rm \
    --name elasticsearch \
    -m 4GB \
    -p 9200:9200 \
    -p 9300:9300 \
    -e "discovery.type=single-node" \
    -e "xpack.security.enabled=false" \
    docker.elastic.co/elasticsearch/elasticsearch:8.4.3
```

In [76]:
# Get df here для удобного написания индекса: 
df

chunk_id article_title  \
0   92dd2f6b5d033c12efb2f8ee3047601e     Abduction   
1   a3b5103db83e97338eb7270a86a63c65     Abduction   
2   81e573a9d329de10a9651c19c4907dc2     Abduction   
3   1738ba0482fa09f893136ec9e10ebff5     Abduction   
4   34bf01c649b729cd3d740f4d5a7cdadf     Abduction   
5   501b7b05cb58454829b17faa1ca0b67f     Abduction   
6   6c943762de83a112fe91d3de929913e8     Abduction   
7   806563744b1ad2c6fdc20c84694a9061     Abduction   
8   2e6cd6b7abb9bc8faf2847668c92552a     Abduction   
9   fb65974393a7620588b7c530d53e9082     Abduction   
10  8499745946def21e570ae45e9a8b2a82     Abduction   
11  1a750c640ab45d4dafc37540fd941b76     Abduction   
12  2bd101c6a5154c559d90b59afe196a2e     Abduction   
13  4f3f89d39f6e17a689d244cf3c1f8db4     Abduction   
14  1c4c184c0ddd0f96760e6fb93aa319fc     Abduction   
15  f7ecd1a32026e385eb2478522b11f166     Abduction   
16  10df8e48bc88212af4767ab5bb2cc362     Abduction   
17  169088bc3b24a5b29621bf52c9a4b680     Abduction   
18  7b955e4553c4bf8f4ff01e2f394ae4b5     Abduction   
19  a769e7233c923d5482ef84a18d34e322     Abduction   
20  c8927e679629bd82151236f34dcb5f8e     Abduction   
21  209998281eac41cd3e3b55194219adaa     Abduction   
22  bd8a142a8ae7d7250e4723f684d50b64     Abduction   
23  bb0345399a3ea255e6d31f230e818d53     Abduction   
24  c9249aa1fd1ac7dc1b7b71164a3de79d     Abduction   
25  bb09ab0b5a7f2dd68c1057691287d985     Abduction   
26  6ae1afd1662837a0294f1ac43b46c62c     Abduction   
27  69da7307a5a7eb89d8c73a877841e42e     Abduction   
28  603011c48c8bfc490ae806a1ba1b143c     Abduction   
29  9f6b5d5c458131ee8547cc2b36012efd     Abduction   
30  e429fb5c3e2a99a629105f9237b66a2d     Abduction   
31  d9e78e99955c0b0ae1dc121c3099336d     Abduction   
32  489106242e16717e1276835ca38b25f9     Abduction   
33  025cb383324ec21cd8cb553cf1a8f8f9     Abduction   
34  17ed98678b554b569ba01d9dce117a44     Abduction   
35  944b677a84a379e8805518f72011e74c     Abduction   
36  5238179758572338829937c7dda59f32     Abduction   
37  d5e38b1922a893a6d1b887c9ea7a6c27     Abduction   
38  5d84563da86fea01adfe1235cd360bdd     Abduction   
39  337d80c564b6a3ca35e08b80301661b8     Abduction   
40  2534a5a8ae83a361431b01c8e721d944     Abduction   

                                         section_path    chunk_type  \
0                                                  []  full_article   
1                    [1. Abduction: The General Idea]       section   
2                    [1. Abduction: The General Idea]     paragraph   
3                    [1. Abduction: The General Idea]     paragraph   
4                    [1. Abduction: The General Idea]     paragraph   
5                    [1. Abduction: The General Idea]     paragraph   
6   [1. Abduction: The General Idea, 1.1 Deduction...    subsection   
7   [1. Abduction: The General Idea, 1.2 The ubiqu...    subsection   
8                          [2. Explicating Abduction]       section   
9                          [2. Explicating Abduction]     paragraph   
10                         [2. Explicating Abduction]     paragraph   
11                         [2. Explicating Abduction]     paragraph   
12                         [2. Explicating Abduction]     paragraph   
13                         [2. Explicating Abduction]     paragraph   
14                         [2. Explicating Abduction]     paragraph   
15                         [2. Explicating Abduction]     paragraph   
16                         [2. Explicating Abduction]     paragraph   
17                         [2. Explicating Abduction]     paragraph   
18                         [2. Explicating Abduction]     paragraph   
19                         [2. Explicating Abduction]     paragraph   
20                         [2. Explicating Abduction]     paragraph   
21                         [2. Explicating Abduction]     paragraph   
22                         [2. Explicating Abduction]     paragraph   
23                         [2. Explicating Abduction]  

In [77]:
# Health check
es_client = Elasticsearch("http://127.0.0.1:9200")
es_client.info()

ObjectApiResponse({'name': '1a9e7e5c2da2', 'cluster_name': 'docker-cluster', 'cluster_uuid': 'LhZCo8veS3CHBYXtwqZ-iQ', 'version': {'number': '8.4.3', 'build_flavor': 'default', 'build_type': 'docker', 'build_hash': '42f05b9372a9a4a470db3b52817899b99a76ee73', 'build_date': '2022-10-04T07:17:24.662462378Z', 'build_snapshot': False, 'lucene_version': '9.3.0', 'minimum_wire_compatibility_version': '7.17.0', 'minimum_index_compatibility_version': '7.0.0'}, 'tagline': 'You Know, for Search'})

In [88]:
from elasticsearch import Elasticsearch, helpers
from tqdm.auto import tqdm

class ElasticRetrieval:
    def __init__(self, es_host: str = "localhost", es_port: int = 9200):
        self.es_client = Elasticsearch([{'host': es_host, 'port': es_port, 'scheme': 'http'}])
        self.es_client.info() #Health check
        self.index_name = "philosophy"
        self.setup_elasticsearch()

    def setup_elasticsearch(self):
        # Define the index mapping
        index_settings = {
            "settings": {
                "number_of_shards": 1,
                "number_of_replicas": 0
            },
            "mappings": {
                "properties": {
                    "chunk_id": {"type": "keyword"},
                    "article_title": {"type": "text"},
                    "section_path": {"type": "keyword"},
                    "chunk_type": {"type": "keyword"},
                    "philosophers": {"type": "keyword"},
                    "concepts": {"type": "keyword"},
                    "content": {"type": "text"},
                    #"vector": {
                    #    "type": "dense_vector",
                    #    "dims": 384  # For all-MiniLM-L6-v2
                    #}
                }
            }
        }
        # Create the index if it doesn't exist
        if not self.es_client.indices.exists(index=self.index_name):
            self.es_client.indices.create(index=self.index_name, body=index_settings)

    def index_chunks(self, chunks: List[Dict[str, Any]]):
        actions = [
            {
                "_index": self.index_name,
                "_id": chunk['chunk_id'],
                "_source": chunk
            }
            for chunk in chunks
        ]
        
        helpers.bulk(self.es_client, actions)

er = ElasticRetrieval()
  
# Ingest articles. Make chunks for 500 (to save time) articles
for article in tqdm(dataset['train'].select(range(500)), desc="Processing articles"):
     chunks = processor.process_article(article)
     er.index_chunks(chunks)

Processing articles:   0%|          | 0/500 [00:00<?, ?it/s]

KeyboardInterrupt: 

In [90]:
## 5. Test Elasticsearch Connection and Query (Optional)

# Test that we can query Elasticsearch
# Simple query to get one document
response = er.es_client.search(
    index="philosophy",
    body={
        "query": {
            "match_all": {}
        },
        "size": 1
    }
)

print("First document in Elasticsearch:")
print(json.dumps(response['hits']['hits'][2]['_source'], indent=2))

First document in Elasticsearch:


IndexError: list index out of range